apps/test/unit/code-studio/pd/workshop_enrollment/enroll_formTest.js

```diff
@@ -55,11 +105,29 @@ describe('Enroll Form', () => {
       enrollForm.setState({role: 'Librarian'});
       refute(enrollForm.exists('#describe_role'));
     });
+
+    extraRequiredParams.forEach(requiredParam => {
+      it(`does not submit when ${requiredParam} is missing`, () => {
+        testValidateFields(
+          enrollForm,
+          omit(requiredParams, requiredParam),
+          requiredParam
+        );
+      });
+    });
+
+    it('submits when all required params are present', () => {
+      testSuccessfulSubmit(enrollForm, requiredParams);
+    });
```

Comment:
This block is added for every form configuration we have –– not sure if there's a cleaner way to do this? It feels like a lot of repeated code, but it also might keep it clear what we're testing and what is failing.

Reply:
Thank you for looking! I believe I did that with `testSuccessfulSubmit` and `ensureValidatedFields`? Are you thinking of something else?

Reply:
I was wondering if it was possible to group the entire 
`extraRequiredParams.forEach(requiredParam => {
      it(`does not submit when ${requiredParam} is missing`, () => {
        testValidateFields(
          enrollForm,
          omit(requiredParams, requiredParam),
          requiredParam
        );
      });
    });

    it('submits when all required params are present', () => {
      testSuccessfulSubmit(enrollForm, requiredParams);
    });`

together?

Reply:
Ah yes, I tried that! But I just re-tried and figured out why it wasn't working –– the `enrollForm` is defined in the `beforeEach` block, which only applies to the `it` methods defined inside that block. So a `describe` block will still have `enroll` form as `undefined`. But if we create a "dummy" `it` block that calls the other functions:
```
it('field validation', () => {
  testRequiredFieldsValidation();
});

const testRequiredFieldsValidation = () => {
  it('first test')

  it('second test')
}
```

The "first test" and "second test" won't get run because they're nested inside of another `it` block.

I did read https://kentcdodds.com/blog/avoid-nesting-when-youre-testing which argues against the practice I see over and over in this code base –– creating undefined variables like `enrollForm` and then using a `beforeEach` block to actually define them. His argument is rather compelling imo. He also talks about React Testing Library which we also don't use ... someday we'll have better react testing norms 🙈

Reply:
Ahhh gotcha! Okay then this all looks great! Some day :)

Reply:
I like what you have here! the top priority for readability in test files is that you can clearly see what the behavior is that's being tested. the helper methods you are using to remove repetition are also good.

one idea which is optional is that you could consider parameterizing the repeated parts of your tests in a loop like this: https://github.com/code-dot-org/code-dot-org/blob/a71f81ba8002621a04df954afda89bcc2b7e4053/apps/test/unit/code-studio/pd/form_components_func/FormControllerTest.js#L429-L432

I would consider this to be about the most abstraction I want to read through when reading unit tests -- it is still very clear what is being tested. however if your `it` methods are not super identical / parameterizable between form types then this may not work, or may work but create more confusion than it saves.